### ISSUE TITLE
Yolo OBB, poor orientation on squares but excellent on rectangles

### ISSUE THREAD

**Issue by bbellocq** [ultralytics/ultralytics#13985]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, 

First of all, I'd like to thank you for what you do. You make life easier for many developers when it comes to object detection. 

For several weeks now, I've been playing with OBBs in YoloV8 and I've noticed that most of my square objects are well sorted, but the bounding boxes have no orientation. I should point out that in the annotation process, the bounding boxes were annotated with an angle and this works perfectly with rectangular shapes. 

I followed the advice of this issue, manually increasing my dataset by playing with the angles. Rectangles are better defined, but squares are not: https://github.com/ultralytics/ultralytics/issues/11676

Are there any new strategies to employ? Do you have a code to modify the box_loss ?

Many thanks to the team and the community. 

I've attached a few photos :
![good_orientation](https://github.com/ultralytics/ultralytics/assets/103435272/789417f0-02b8-4506-b750-4676df97a2e4)
![bad_orientation](https://github.com/ultralytics/ultralytics/assets/103435272/f7b51d98-f7c6-45c5-9083-66044fb37cad)

**Comment by glenn-jocher** [comment]
@bbellocq hello,

Thank you for your kind words and for sharing your experience with YOLOv8's Oriented Bounding Boxes (OBBs). It's great to hear that the tool has been beneficial for your object detection tasks.

Regarding the issue you're encountering with square objects not having the correct orientation, this is a known challenge in OBB detection. Squares, due to their symmetry, can sometimes confuse the model when it comes to determining orientation. Here are a few strategies you might consider to improve the orientation accuracy for square objects:

1. **Data Augmentation**: While you've already increased your dataset by varying angles, consider adding more diverse augmentations. Techniques like rotation, scaling, and flipping can help the model learn better orientation cues.

2. **Loss Function Adjustment**: Modifying the box loss function can sometimes help. You can experiment with different loss components or weights to emphasize orientation accuracy. Here's a basic example of how you might adjust the loss function in YOLOv8:

    ```python
    from ultralytics import YOLO

    # Load your model
    model = YOLO("yolov8n-obb.yaml")

    # Custom training with modified loss function
    def custom_loss(pred, target):
        # Implement your custom loss logic here
        pass

    model.train(data="DOTAv1.yaml", epochs=100, imgsz=640, loss_fn=custom_loss)
    ```

3. **Model Architecture**: Sometimes, tweaking the model architecture can yield better results. Experiment with different backbone networks or additional layers that might help the model capture orientation features more effectively.

4. **Hyperparameter Tuning**: Adjusting hyperparameters such as learning rate, batch size, and optimizer settings can also impact the model's performance on specific tasks like orientation detection.

5. **Annotation Quality**: Ensure that your annotations are as precise as possible. Any inconsistencies in the annotation process can lead to poor model performance.

If you haven't already, please ensure you're using the latest version of the Ultralytics YOLO package, as updates often include performance improvements and bug fixes.

For more detailed guidance on working with OBB datasets, you can refer to our documentation: [Oriented Bounding Box (OBB) Datasets Overview](https://docs.ultralytics.com/datasets/obb/).

Feel free to share any additional details or code snippets that might help us better understand your setup. The community and the Ultralytics team are here to support you!

**Comment by bbellocq** [comment]
Hello @glenn-jocher, 

Thanks for your reply! I was wondering, does calling the loss_fn argument work during training? I can't find this argument in the documentation : https://docs.ultralytics.com/modes/train/#logging

Thanks a lot.

**Comment by glenn-jocher** [comment]
Hello @bbellocq,

Thank you for your follow-up! You are correct; the `loss_fn` argument is not currently documented in the official Ultralytics YOLO training documentation. My apologies for any confusion.

To customize the loss function, you would typically need to modify the training script directly. Here's a more detailed approach:

1. **Custom Training Script**: You can create a custom training script where you define and use your custom loss function. Here's a basic example:

    ```python
    from ultralytics import YOLO
    import torch

    # Define your custom loss function
    def custom_loss(pred, target):
        # Implement your custom loss logic here
        loss = torch.nn.functional.mse_loss(pred, target)
        return loss

    # Load your model
    model = YOLO("yolov8n-obb.yaml")

    # Custom training loop
    for epoch in range(100):  # example for 100 epochs
        for batch in dataloader:  # assuming you have a dataloader
            images, targets = batch
            preds = model(images)
            loss = custom_loss(preds, targets)
            loss.backward()
            optimizer.step()
            optimizer.zero_grad()
    ```

2. **Verify with Latest Version**: Ensure you are using the latest version of the Ultralytics YOLO package to benefit from recent updates and bug fixes. You can update your package using:
    ```bash
    pip install --upgrade ultralytics
    ```

3. **Reproducible Example**: If you encounter any issues, providing a minimum reproducible example can significantly help in diagnosing the problem. You can find more details on creating one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to share any specific issues or errors you encounter, and we can assist you further. The community and the Ultralytics team are here to help! 😊

**Comment by Viral97** [comment]
The issue still persist

**Comment by glenn-jocher** [comment]
Thank you for pointing that out! To address this, you’d need to modify the loss function directly in the YOLO source code, typically in the `ultralytics/utils/loss.py` file. Refer to the [loss function documentation](https://docs.ultralytics.com/reference/utils/loss/) for more details on customizing loss calculations. Let us know if you need further clarification!

**Comment by Petros626** [comment]
Note for me

**Comment by glenn-jocher** [comment]
@Viral97 thanks for the update. To help troubleshoot the square orientation issue, could you share some specifics about your setup - which YOLO11 OBB model you're using, your dataset details, and any training configurations you've tried? 

For square objects, the fundamental challenge is their rotational symmetry makes orientation ambiguous to the model. You might also try training with the latest YOLO11 OBB models which have improved architecture, or experiment with different annotation strategies where you consistently annotate squares with a specific reference edge (like always starting from the top edge) to give the model a clearer orientation signal.

**Comment by Scherlac** [comment]
Hi @glenn-jocher,

I did some evaluation on training YOLO11 OBB on shapes (square, rectangle, circle, ellipsis) with 100% synthetic data. Based on label annotation the images and labels are correct:

![labels vs images](https://github.com/Scherlac/yolo-shapes/raw/refs/heads/feature/train_yolo11-obb/runs/obb/yolo_obb_shapes_training-square-issue/val_batch0_labels.jpg)

I've also limited the rotation angle for squares (pi/4 ... p/2] and for circles (zero) to avoid issues related to symmetry.

While creating the labels with the build in [xywhr2xyxyxyxy()](https://github.com/ultralytics/ultralytics/blame/main/ultralytics/utils/ops.py#L361) the training results are looking good for other shapes, the orientations of squares are quite poor:

![poor orientation of squares](https://github.com/Scherlac/yolo-shapes/raw/refs/heads/feature/train_yolo11-obb/runs/obb/yolo_obb_shapes_training-square-issue/val_batch0_pred.jpg)

Interestingly after I choose to label the left half of the squares (rotated pi/4... pi/2, hence looks like top half) the training went unexpectedly well:

![correct orientation of squares](https://media.githubusercontent.com/media/Scherlac/yolo-shapes/refs/heads/feature/train_yolo11-obb/runs/obb/yolo_obb_shapes_training-with-hack/val_batch0_pred.jpg)


Regarding to rotation topic, why the top half is shown selected on images:

Standard coordinate system (→, ↑, geometry, center origo) the rotation using sin, cos functions with positive angle will result ccw (↺) rotation.
Computer graphics coordinate system (→, ↓, due to upside-down `y`, with top left origo) the rotation using sin, cos function with positive angle will result cw (↻) rotation. 


Edit:
I saw the discussion on the `regularize_rboxes` at [OBB Datasets Overview](https://docs.ultralytics.com/datasets/obb/). Can we disable the h <--> w swap somehow? For squares the `h` and `w` are approximately equal, and it will swap 50% of the case, possibly causing the square related issue. Since we use cv2's minAreaRect the swap shouldn't be needed!?

Edit2:
I've checked and the `regularization_rboxes` is not in use during training so it can not affect this issue.

Edit3:
The following line causing probably the issue [link](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/metrics.py#L206):

```python
    return a * cos2 + b * sin2, a * sin2 + b * cos2, (a - b) * cos * sin
```

This term is insensitive to $$Theta$$ in case `a` and `b` gets close to each other. After some fix all squares are trained correctly as well! 

I'm not sure if only this implementation is incorrect or the original paper has some flaw. 

Edit4:

Yup, the demonstration implementation linked to the paper also has this issue [link](https://github.com/ProbIOU/probiou-sample/blob/main/probiou_pytorch.py#L11):

```python
def gbb_form(boxes):
    return torch.cat((boxes[:,:2],torch.pow(boxes[:,2:4],2)/12,boxes[:,4:]),1)

def rotated_form(a_, b_, angles):

    a   = a_*torch.pow(torch.cos(angles),2.)+b_*torch.pow(torch.sin(angles),2.)
    b   = a_*torch.pow(torch.sin(angles),2.)+b_*torch.pow(torch.cos(angles),2.)
    c   = a_*torch.cos(angles)*torch.sin(angles)-b_*torch.sin(angles)*torch.cos(angles)
    return a,b,c

def probiou_loss(pred, target, eps = 1e-3, mode='l1'):

    """
        pred    -> a matrix [N,5](x,y,w,h,angle - in radians) containing ours predicted box ;in case of HBB angle == 0
        target  -> a matrix [N,5](x,y,w,h,angle - in radians) containing ours target    box ;in case of HBB angle == 0
        eps     -> threshold to avoid infinite values
        mode    -> ('l1' in [0,1] or 'l2' in [0,inf]) metrics according our paper

    """

    gbboxes1 = gbb_form(pred)
    gbboxes2 = gbb_form(target)

    x1, y1, a1_, b1_, c1_ = gbboxes1[:,0], gbboxes1[:,1], gbboxes1[:,2], gbboxes1[:,3], gbboxes1[:,4]
    x2, y2, a2_, b2_, c2_ = gbboxes2[:,0], gbboxes2[:,1], gbboxes2[:,2], gbboxes2[:,3], gbboxes2[:,4]

    a1, b1, c1 = rotated_form(a1_, b1_, c1_)
    a2, b2, c2 = rotated_form(a2_, b2_, c2_)
```

[isotropic issues on square OBB](https://github.com/ProbIOU/probiou-sample/issues/2)

**Comment by glenn-jocher** [comment]
@Scherlac this matches a fundamental limitation: for near-isotropic boxes (`w≈h`) the OBB parameterization becomes angle-ambiguous, and any overlap-based objective (including `probiou()` used in the rotated loss path) becomes nearly angle-invariant, so gradients on `pred_angle` can collapse for true squares (your “half-square” tweak works because it breaks symmetry). The `regularize_rboxes()` swap isn’t part of the training-loss path and the 0–90° angle constraint is a core YOLO OBB assumption, so there’s no supported switch to “disable swapping” without redesigning the representation end-to-end; see the [OBB task notes on angle range](https://docs.ultralytics.com/tasks/obb/) and the [loss/`probiou` references](https://docs.ultralytics.com/reference/utils/loss/). If you have a minimal repro + a patch that improves square angle learning without regressing other shapes, please open a new issue (this one is stale/closed) and include:

```bash
yolo checks
```

**Comment by Scherlac** [comment]
@glenn-jocher 

It has nothing to do with the symmetry of the shape, is about the aspect ratio of the BB. Yes if `w≈h`, and that leads to the solution.  Before the `a1, b1, a2, b2` is calculated we can introduce a random distortion in each iteration. But is has to be random, otherwise some other aspect ratio would suffer from the same issue.
   
Below shown the result after fix was applied:

![result with fixed probiou](https://media.githubusercontent.com/media/Scherlac/yolo-shapes/refs/heads/feature/train_yolo11-obb/runs/obb/yolo_obb_shapes_training/val_batch0_pred.jpg)

**Comment by Scherlac** [comment]
@glenn-jocher 

I've created a new issue #23011 and a PR #23012

**Comment by glenn-jocher** [comment]
@Scherlac thanks for opening the new report and PR (#23011 / #23012) — please keep the change tightly scoped to the `probiou()` / `_get_covariance_matrix()` path and add a small regression test (or minimal script + expected output) that shows square angle learning improves without hurting rectangles/typical OBB cases; we’ll continue review/discussion on the new issue/PR (no ETA on merge).

**Comment by Y-T-G** [comment]
You can try YOLO26-OBB. There's a new angle loss.
